src/server/frontend/wayland/wayland_connector.cpp

```diff
@@ -1954,6 +1961,532 @@ class WlShell : public wayland::Shell
+            }));
+    }
+
+    std::shared_ptr<bool> const destroyed;
+};
+
+struct XdgSurfaceV6 : wayland::XdgSurfaceV6, WlAbstractMirWindow
+{
+    XdgSurfaceV6* get_xdgsurface(wl_resource* surface) const
+    {
+        auto* tmp = wl_resource_get_user_data(surface);
+        return static_cast<XdgSurfaceV6*>(static_cast<wayland::XdgSurfaceV6*>(tmp));
+    }
+
+    XdgSurfaceV6(wl_client* client,
+        wl_resource* parent,
+        uint32_t id,
+        wl_resource* surface,
+        std::shared_ptr<mf::Shell> const& shell,
+        WlSeat& seat) :
+        wayland::XdgSurfaceV6(client, parent, id),
+        WlAbstractMirWindow{client, surface, resource, shell},
+        parent{parent},
+        shell{shell},
+        sink{std::make_shared<XdgSurfaceV6EventSink>(&seat, client, surface, resource, destroyed)}
+    {
+        WlAbstractMirWindow::sink = sink;
+    }
+
+    ~XdgSurfaceV6() override
+    {
+        auto* const mir_surface = get_wlsurface(surface);
+        mir_surface->set_role(&nullwlmirwindow);
+    }
+
+    void destroy() override
+    {
+        wl_resource_destroy(resource);
+    }
+
+    void get_toplevel(uint32_t id) override
+    {
+        new XdgToplevelV6{client, parent, id, shell, this};
+        auto* const mir_surface = get_wlsurface(surface);
+        mir_surface->set_role(this);
+    }
+
+    void get_popup(uint32_t id, struct wl_resource* parent, struct wl_resource* positioner) override
+    {
+        auto* tmp = wl_resource_get_user_data(positioner);
+        auto const* const pos =  static_cast<XdgPositionerV6*>(static_cast<wayland::XdgPositionerV6*>(tmp));
+
+        auto const session = session_for_client(client);
+        auto& parent_surface = *get_xdgsurface(parent);
+
+        params.type = mir_window_type_freestyle;
+        params.parent_id = parent_surface.surface_id;
+        if (pos->size.is_set()) params.size = pos->size.value();
+        params.aux_rect = pos->aux_rect;
+        params.surface_placement_gravity = pos->surface_placement_gravity;
+        params.aux_rect_placement_gravity = pos->aux_rect_placement_gravity;
+        params.aux_rect_placement_offset_x = pos->aux_rect_placement_offset_x;
+        params.aux_rect_placement_offset_y = pos->aux_rect_placement_offset_y;
+        params.placement_hints = mir_placement_hints_slide_any;
+
+        new XdgPopupV6{client, parent, id};
+        auto* const mir_surface = get_wlsurface(surface);
+        mir_surface->set_role(this);
+    }
+
+    void set_window_geometry(int32_t x, int32_t y, int32_t width, int32_t height) override
+    {
+        geom::Rectangle const input_region{{x, y}, {width, height}};
+
+        if (surface_id.as_value())
+        {
+            auto const session = session_for_client(client);
+            auto const surface = get_surface_for_id(session, surface_id);
+
+            shell::SurfaceSpecification modifications;
+            modifications.input_shape = {input_region};
```

Comment:
set_window_geometry maps to Mir's input shaping?

Reply:
That's my reading of the documentation